server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/MailLoader.java

```diff
@@ -38,18 +40,26 @@
         this.blobIdFactory = blobIdFactory;
     }
 
-    MailWithEnqueueId load(MailReferenceDTO dto) throws MailQueue.MailQueueException {
-        try {
-            MailReference mailReference = dto.toMailReference(blobIdFactory);
+    Mono<MailWithEnqueueId> load(MailReferenceDTO dto) {
+        return Mono.fromCallable(() -> dto.toMailReference(blobIdFactory))
+            .flatMap(mailReference -> buildMail(mailReference)
+                .map(mail -> new MailWithEnqueueId(mailReference.getEnqueueId(), mail)));
+    }
 
-            Mail mail = mailReference.getMail();
-            MimeMessage mimeMessage = mimeMessageStore.read(mailReference.getPartsId()).block();
+    private Mono<Mail> buildMail(MailReference mailReference) {
+        return mimeMessageStore.read(mailReference.getPartsId())
+            .flatMap(mimeMessage -> buildMailWithMessageReference(mailReference, mimeMessage));
+    }
+
+    private Mono<Mail> buildMailWithMessageReference(MailReference mailReference, MimeMessage mimeMessage) {
+        Mail mail = mailReference.getMail();
+        try {
             mail.setMessage(mimeMessage);
```

Comment:
Looking at the complexity of MailImpl::setMessage I'd argue it can take time and it can fail so maybe we can benefit from wrapping it in a ractive operation.